src/main/java/com/google/api/codegen/configgen/CollectionPattern.java

```diff
@@ -34,45 +38,136 @@ public static CollectionPattern create(FieldSegment fieldSegment) {
+   * @param index the index of the segment in segments
+   * @param pathSegments the full path containing segment at index
+   * @return a templatized path segment with any wildcard segment replaced with a literal segment
+   *     containing the entity name
+   */
+  private static PathSegment buildTemplatizedSegment(
+      PathSegment segment, int index, List<PathSegment> pathSegments) {
+    if (!(segment instanceof WildcardSegment)) {
+      return segment;
+    }
 
-  private CollectionPattern(FieldSegment fieldSegment) {
-    this.fieldSegment = fieldSegment;
+    String prefix = buildSimpleEntityPrefix(pathSegments, index);
+    String entityName = buildEntityName(pathSegments, index, prefix);
+    if (((WildcardSegment) segment).isUnbounded()) {
+      return new LiteralSegment(String.format("{%s=**}", entityName));
+    }
 
-    ImmutableList.Builder<PathSegment> builder = ImmutableList.<PathSegment>builder();
-    PathSegment lastSegment = null;
-    String simpleName = null;
-    for (PathSegment pathSegment : fieldSegment.getSubPath()) {
-      if (pathSegment instanceof WildcardSegment) {
-        String name = "unknown";
-        String suffix = "";
-        if (lastSegment != null && lastSegment instanceof LiteralSegment) {
-          name =
-              LanguageUtil.upperCamelToLowerUnderscore(Inflector.singularize(lastSegment.syntax()));
-        }
-        if (((WildcardSegment) pathSegment).isUnbounded()) {
-          name += "_path";
-          suffix = "=**";
-        }
-        builder.add(new LiteralSegment(String.format("{%s%s}", name, suffix)));
-        // simpleName will be set to the name of the last wildcard to be templatized
-        simpleName = name;
-      } else {
-        builder.add(pathSegment);
-      }
-      lastSegment = pathSegment;
+    return new LiteralSegment(String.format("{%s}", entityName));
+  }
+
+  /**
+   * Builds an entity name for a wildcard path segment, handling singular resources.
+   *
+   * <p>The entity name of a singular resource will be prefixed by its parent collection resource.
+   *
+   * @param pathSegments the full path containing the wildcard segment at wildcardIndex
+   * @param wildcardIndex the index of the wildcard segment
+   * @return the entity name that corresponds to the wildcard segment
+   */
+  private static String buildEntityName(List<PathSegment> pathSegments, int wildcardIndex) {
+    String prefix = buildEntityPrefix(pathSegments, wildcardIndex);
+    return buildEntityName(pathSegments, wildcardIndex, prefix);
+  }
+
+  /**
+   * Builds an entity name for a wildcard path segment.
+   *
+   * @param pathSegments the full path containing the wildcard segment at wildcardIndex
+   * @param wildcardIndex the index of the wildcard segment
+   * @param prefix the prefix to the entity name. Should be the resource name.
+   * @return the entity name that corresponds to the wildcard segment
+   */
+  private static String buildEntityName(
```

Comment:
I found it quite hard to parse this section of code, because there are multiple functions with the same name that call each other. I have suggested some renamings that might clarify things a bit, let me know what you think.

Suggested renaming: `appendUnboundWildcardSuffix`

Reply:
Done